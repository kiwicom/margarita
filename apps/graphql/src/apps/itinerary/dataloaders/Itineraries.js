// @flow

import stringify from 'json-stable-stringify';
import qs from 'querystring';
import * as DateFNS from 'date-fns';
import { OptimisticDataloader } from '@kiwicom/graphql-utils';
import { UK_DATE_FORMAT } from '@kiwicom/margarita-config';

import fetch from '../../../services/fetch/tequilaFetch';
import {
  mapLocation,
  mapDate,
  getItineraryType,
  mapSectors,
} from '../helpers/Itineraries';
import type {
  ItinerariesReturnSearchParameters,
  ItinerariesOneWaySearchParameters,
  ApiResponseType,
  Itinerary,
} from '../Itinerary';

const stripTimeZoneOffset = (date: Date) =>
  DateFNS.addMinutes(date, date.getTimezoneOffset());

const parseDate = (date: Date) =>
  DateFNS.format(stripTimeZoneOffset(date), UK_DATE_FORMAT);

export const parseParameters = (
  input: ItinerariesReturnSearchParameters | ItinerariesOneWaySearchParameters,
) => {
  const { origin, destination, outboundDate } = input.itinerary;
  const inboundDate = input.itinerary.inboundDate
    ? input.itinerary.inboundDate
    : null;

  const flyFrom = origin.ids.join();
  const flyTo = destination && destination.ids ? destination.ids.join() : null;

  const commonSearchParams = {
    fly_from: flyFrom,
    ...(input.order && { asc: input.order === 'ASC' ? 1 : 0 }),
    ...(input.sort && { sort: input.sort }),
    date_from: parseDate(outboundDate.start),
    date_to: outboundDate.end ? parseDate(outboundDate.end) : null,
    fly_to: flyTo,
    ...(input.passengers && {
      adults: input.passengers.adults ?? 0,
      children: input.passengers.children ?? 0,
      infants: input.passengers.infants ?? 0,
    }),
    curr: 'EUR',
  };

  return {
    ...commonSearchParams,
    ...addReturnSearchQueryParams(inboundDate),
  };
};

const addReturnSearchQueryParams = inboundDate => {
  return {
    ...(inboundDate && {
      return_from: parseDate(inboundDate.start),
      ...(inboundDate.end && {
        return_to: parseDate(inboundDate.end),
      }),
    }),
  };
};

const fetchItineraries = async (
  parameters: $ReadOnlyArray<ItinerariesReturnSearchParameters>,
) => {
  const results: $ReadOnlyArray<ApiResponseType> = await Promise.all(
    parameters.map(params => {
      return fetch(`/v2/search?${qs.stringify(parseParameters(params))}`);
    }),
  );
  return results.map(res => {
    return sanitizeItineraries(res);
  });
};

const sanitizeItineraries = (response: ApiResponseType): Itinerary[] => {
  const itineraries = response.data;

  return itineraries.map(itinerary => {
    return {
      id: itinerary.id,
      type: getItineraryType(itinerary.routes),
      bookingToken: itinerary.booking_token,
      isValid: false,
      isChecked: false,
      startTime: mapDate(itinerary.local_departure, itinerary.utc_departure),
      endTime: mapDate(itinerary.local_arrival, itinerary.utc_arrival),
      destination: mapLocation(
        itinerary.flyTo,
        itinerary.cityTo,
        itinerary.countryTo?.name,
        itinerary.countryTo?.code,
      ),
      origin: mapLocation(
        itinerary.flyFrom,
        itinerary.cityFrom,
        itinerary.countryFrom?.name,
        itinerary.countryFrom?.code,
      ),
      sectors: mapSectors(itinerary.route, itinerary.routes),
      price: {
        currency: response.currency,
        amount: itinerary.price,
      },
    };
  });
};

export function createItinerariesLoader() {
  return new OptimisticDataloader(
    async (
      keys: $ReadOnlyArray<ItinerariesReturnSearchParameters>,
    ): Promise<Array<Itinerary[] | Error>> => fetchItineraries(keys),
    {
      cacheKeyFn: stringify,
    },
  );
}
