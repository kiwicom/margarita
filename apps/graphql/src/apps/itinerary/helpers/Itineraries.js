// @flow

import * as DateFNS from 'date-fns';
import { head, last } from 'ramda';
import { fromGlobalId } from '@kiwicom/graphql-global-id';

import type { ApiRouteItem, Sector, Segment } from '../Itinerary';
import type { RouteStop } from '../../booking/Booking';

export const differenceInMinutes = (
  from: ?(string | number),
  to: ?(string | number),
) => {
  if (from == null || to == null) {
    return null;
  }
  const parseNonSpecificDate = date =>
    typeof date === 'string'
      ? DateFNS.parseISO(date)
      : DateFNS.fromUnixTime(date);

  return parseInt(
    DateFNS.differenceInMinutes(
      parseNonSpecificDate(to),
      parseNonSpecificDate(from),
    ),
    10,
  );
};

export const mapToOpaqueID = (input: any) => {
  const parsedOriginIds: string[] = input.itinerary.origin.ids.map(id =>
    fromGlobalId(id),
  );
  const parsedDestinationIds: string[] = input.itinerary.destination.ids.map(
    id => fromGlobalId(id),
  );

  return {
    ...input,
    itinerary: {
      ...input.itinerary,
      origin: {
        ids: parsedOriginIds,
      },
      destination: {
        ids: parsedDestinationIds,
      },
    },
  };
};

export const mapVehicle = (type: ?string, uniqueNo: ?string) => ({
  type: type ?? null,
  uniqueNo: uniqueNo ?? null,
});

export const mapTransporter = (transporter: ?string) => ({
  name: transporter ?? null,
});

export const getItineraryType = (routes: ?Array<Array<string>>) => {
  if (routes == null) {
    return null;
  }
  if (routes.length === 1) {
    return 'oneway';
  }
  if (routes.length === 2) {
    return 'return';
  }
  return null;
};

/**
 * This function converts route and routes data from REST API to well-structured Sectors with Segments.
 * Sector = The whole part from source to destination, e.g Oslo -> Prague. A return trip will have 2 sectors
 * and 1 way trip will have 1 sector. A multicity trip could have 2 or more sectors.
 * Segment = A part of the sector. Segments could be Oslo -> Warzaw, Warzaw -> Prague
 */
export const mapSectors = (
  routesList: ?(ApiRouteItem[]),
  routeCodes: ?Array<string[]>,
): ?Array<Sector> => {
  if (routesList == null || routeCodes == null) {
    return null;
  }
  const sectors = [];
  let currentSectorIndex = 0;
  let currentRouteCode = head(routeCodes) ?? [];

  let { currentSector, currentArrival } = sanitizeSector(
    routesList,
    currentRouteCode,
  );

  sectors[currentSectorIndex] = currentSector;

  routesList.forEach((route: ApiRouteItem) => {
    const segment = sanitizeSegment(route);

    if (sectors[currentSectorIndex].segments) {
      sectors[currentSectorIndex].segments.push(segment);
    }

    if (route.flyTo === currentArrival?.flyTo) {
      currentSectorIndex++;
      // $FlowExpectedError: We already tested that routeCodes != null
      currentRouteCode = routeCodes[currentSectorIndex];

      if (currentRouteCode == null) {
        return;
      }
      // $FlowExpectedError: We already tested that routesList != null
      const nextSector = sanitizeSector(routesList, currentRouteCode);
      currentArrival = nextSector.currentArrival;
      sectors[currentSectorIndex] = nextSector.currentSector;
    }
  });

  return sectors;
};

const sanitizeSector = (
  routesList: ApiRouteItem[],
  currentRouteCode: string[],
) => {
  const currentRouteArrivalCode = last(currentRouteCode) ?? '';
  const currentRouteDepartureCode = head(currentRouteCode) ?? '';

  const currentDeparture: ?ApiRouteItem = routesList.find(
    route => route.flyFrom === currentRouteDepartureCode,
  );

  const currentArrival: ?ApiRouteItem = routesList.find(
    route => route.flyTo === currentRouteArrivalCode,
  );

  const departure = sanitizeRouteStop(
    apiRouteItemToDeparture(currentDeparture),
  );
  const arrival = sanitizeRouteStop(apiRouteItemToArrival(currentArrival));

  const currentSector = {
    departure,
    arrival,
    arrivalTime: {
      local: arrival.time?.local,
      utc: arrival.time?.utc,
    },
    departureTime: {
      local: departure.time?.local,
      utc: departure.time?.utc,
    },
    duration: differenceInMinutes(departure.time?.utc, arrival.time?.utc),
    destination: mapLocation(currentArrival?.flyTo, currentArrival?.cityTo),
    origin: mapLocation(currentDeparture?.flyFrom, currentDeparture?.cityFrom),
    segments: [],
    stopoverDuration: 0,
  };

  return {
    currentArrival,
    currentSector,
  };
};

const apiRouteItemToArrival = (routeItem: ?ApiRouteItem) => ({
  cityName: routeItem?.cityTo,
  code: routeItem?.flyTo,
  time: {
    utc: routeItem?.utc_arrival,
    local: routeItem?.local_arrival,
  },
});

const apiRouteItemToDeparture = (routeItem: ?ApiRouteItem) => ({
  cityName: routeItem?.cityFrom,
  code: routeItem?.flyFrom,
  time: {
    utc: routeItem?.utc_departure,
    local: routeItem?.local_departure,
  },
});

const sanitizeRouteStop = ({
  cityName,
  code,
  time,
}: {|
  +cityName: ?string,
  +code: ?string,
  +time: {| +utc: ?string, +local: ?string |},
|}): RouteStop => ({
  cityName,
  cityId: null,
  time,
  code,
});

const sanitizeSegment = (segment: ?ApiRouteItem): Segment => {
  return {
    duration: differenceInMinutes(segment?.utc_departure, segment?.utc_arrival),
    id: segment?.id,
    transporter: mapTransporter(segment?.airline),
    vehicle: mapVehicle(segment?.vehicle_type, String(segment?.flight_no)),
    departure: sanitizeRouteStop(apiRouteItemToDeparture(segment)),
    arrival: sanitizeRouteStop(apiRouteItemToArrival(segment)),
  };
};

export const mapDate = (local: ?string, utc: ?string) => {
  const localDate = local && DateFNS.parseISO(local).toISOString();
  const utcDate = utc && DateFNS.parseISO(utc).toISOString();
  return {
    local: localDate ?? null,
    utc: utcDate ?? null,
  };
};

export const mapLocationArea = (
  code: ?string,
  locationId: ?string,
  name: ?string,
  slug: ?string,
  flagURL: ?string,
) => ({
  id: locationId ?? null,
  code: code ?? null,
  locationId: locationId ?? null,
  name: name ?? null,
  slug: slug ?? null,
  flagURL: flagURL ?? null,
});

export const mapLocation = (
  locationId: ?string,
  name: ?string,
  countryName: ?string,
  countryCode: ?string,
  slug: ?string,
) => ({
  id: locationId ?? null,
  locationId: locationId ?? null,
  name: name ?? null,
  timezone: 'UTC+1',
  country: mapLocationArea(countryCode, countryCode, countryName),
  slug: slug ?? null,
  city: null,
  type: null,
});
