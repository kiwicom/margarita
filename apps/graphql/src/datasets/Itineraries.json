{
  "ref_tasks": [],
  "search_params": {
    "to_type": "airport",
    "flyFrom_type": "airport",
    "seats": {
      "infants": 0,
      "passengers": 1,
      "adults": 1,
      "children": 0
    }
  },
  "currency_rate": 1,
  "refresh": [],
  "connections": [],
  "currency": "EUR",
  "del": null,
  "all_airlines": [],
  "time": 1,
  "all_stopover_airports": [],
  "data": [
    {
      "quality": 994.998825,
      "flyTo": "LIM",
      "deep_link": "https://mock.url",
      "nightsInDest": null,
      "airlines": [
        "ET",
        "B6",
        "VV",
        "DY"
      ],
      "pnr_count": 4,
      "baglimit": {},
      "has_airport_change": false,
      "distance": 11079.54,
      "type_flights": [
        "lcc-VV",
        "lcc-ET",
        "lcc-DY",
        "lcc-B6"
      ],
      "bags_price": {},
      "flyFrom": "OSL",
      "cityFrom": "Oslo",
      "booking_token": "token-1",
      "duration": {
        "total": 206100,
        "return": 0,
        "departure": 206100
      },
      "id": "3991227978220119_0|3915361603311185_0|8776301839864123_0|12345316616561678_0",
      "conversion": {
        "EUR": 525
      },
      "countryTo": {
        "code": "PE",
        "name": "Peru"
      },
      "price": 525,
      "routes": [
        [
          "OSL",
          "LIM"
        ]
      ],
      "cityTo": "Lima",
      "transfers": [],
      "route": [
        {
          "refresh_timestamp": "1970-01-01T00:00:00.000Z",
          "bags_recheck_required": false,
          "return": 0,
          "flight_no": 715,
          "operating_carrier": "ET",
          "cityTo": "Stockholm",
          "vehicle_type": "aircraft",
          "flyFrom": "OSL",
          "id": "3991227978220119_0",
          "equipment": "77L",
          "combination_id": "3991227978220119",
          "fare_family": "",
          "flyTo": "ARN",
          "airline": "ET",
          "fare_classes": "E",
          "cityFrom": "Oslo",
          "last_seen": "2019-02-15T13:18:40.000Z",
          "guarantee": false,
          "fare_basis": "EOWNO",
          "local_arrival": "2019-03-01T19:50:00.000Z",
          "utc_arrival": "2019-03-01T18:50:00.000Z",
          "local_departure": "2019-03-01T18:45:00.000Z",
          "utc_departure": "2019-03-01T17:45:00.000Z"
        },
        {
          "refresh_timestamp": "1970-01-01T00:00:00.000Z",
          "bags_recheck_required": false,
          "return": 0,
          "flight_no": 7035,
          "operating_carrier": null,
          "cityTo": "Fort Lauderdale",
          "vehicle_type": "aircraft",
          "flyFrom": "ARN",
          "id": "3915361603311185_0",
          "equipment": null,
          "combination_id": "3915361603311185",
          "fare_family": "",
          "flyTo": "FLL",
          "airline": "DY",
          "fare_classes": "",
          "cityFrom": "Stockholm",
          "last_seen": "2019-02-15T12:16:02.000Z",
          "guarantee": true,
          "fare_basis": "",
          "local_arrival": "2019-03-02T19:45:00.000Z",
          "utc_arrival": "2019-03-03T00:45:00.000Z",
          "local_departure": "2019-03-02T15:15:00.000Z",
          "utc_departure": "2019-03-02T14:15:00.000Z"
        },
        {
          "refresh_timestamp": "2019-02-15T14:47:37.000Z",
          "bags_recheck_required": false,
          "return": 0,
          "flight_no": 41,
          "operating_carrier": "B6",
          "cityTo": "Medellín",
          "vehicle_type": "aircraft",
          "flyFrom": "FLL",
          "id": "8776301839864123_0",
          "equipment": "320",
          "combination_id": "8776301839864123",
          "fare_family": "",
          "flyTo": "MDE",
          "airline": "B6",
          "fare_classes": "",
          "cityFrom": "Fort Lauderdale",
          "last_seen": "2019-02-15T14:47:37.000Z",
          "guarantee": true,
          "fare_basis": "",
          "local_arrival": "2019-03-03T02:24:00.000Z",
          "utc_arrival": "2019-03-03T07:24:00.000Z",
          "local_departure": "2019-03-02T22:55:00.000Z",
          "utc_departure": "2019-03-03T03:55:00.000Z"
        },
        {
          "refresh_timestamp": "1970-01-01T00:00:00.000Z",
          "bags_recheck_required": false,
          "return": 0,
          "flight_no": 441,
          "operating_carrier": "VV",
          "cityTo": "Lima",
          "vehicle_type": "aircraft",
          "flyFrom": "MDE",
          "id": "12345316616561678_0",
          "equipment": "Non",
          "combination_id": "12345316616561678",
          "fare_family": "",
          "flyTo": "LIM",
          "airline": "VV",
          "fare_classes": "",
          "cityFrom": "Medellín",
          "last_seen": "2019-02-15T12:41:52.000Z",
          "guarantee": true,
          "fare_basis": "",
          "local_arrival": "2019-03-03T22:00:00.000Z",
          "utc_arrival": "2019-03-04T03:00:00.000Z",
          "local_departure": "2019-03-03T18:50:00.000Z",
          "utc_departure": "2019-03-03T23:50:00.000Z"
        }
      ],
      "baggage": {},
      "facilitated_booking_available": false,
      "countryFrom": {
        "code": "NO",
        "name": "Norway"
      },
      "local_arrival": "2019-03-03T22:00:00.000Z",
      "utc_arrival": "2019-03-04T03:00:00.000Z",
      "local_departure": "2019-03-01T18:45:00.000Z",
      "utc_departure": "2019-03-01T17:45:00.000Z"
    },
    {
      "quality": 941.465668,
      "flyTo": "LIM",
      "deep_link": "https://mock.url2",
      "nightsInDest": null,
      "airlines": [
        "VH",
        "DL",
        "NK",
        "DY"
      ],
      "pnr_count": 4,
      "baglimit": {
        "hold_weight": 18
      },
      "has_airport_change": false,
      "distance": 11079.54,
      "type_flights": [
        "GDS-r",
        "lcc-NK",
        "lcc-DY",
        "lcc-VH"
      ],
      "bags_price": {
        "1": 165.65
      },
      "flyFrom": "OSL",
      "booking_token": "token-2",
      "cityFrom": "Oslo",
      "duration": {
        "total": 174360,
        "return": 0,
        "departure": 174360
      },
      "id": "3923058182037843_0|16669696135344215_0|10000058272100155_0|19795607481891218_0",
      "conversion": {
        "EUR": 542
      },
      "countryTo": {
        "code": "PE",
        "name": "Peru"
      },
      "price": 542,
      "routes": [
        [
          "OSL",
          "LIM"
        ]
      ],
      "cityTo": "Lima",
      "transfers": [],
      "route": [
        {
          "refresh_timestamp": "1970-01-01T00:00:00.000Z",
          "bags_recheck_required": false,
          "return": 0,
          "flight_no": 7001,
          "operating_carrier": null,
          "cityTo": "New York",
          "vehicle_type": "aircraft",
          "flyFrom": "OSL",
          "id": "3923058182037843_0",
          "equipment": null,
          "combination_id": "3923058182037843",
          "fare_family": "",
          "flyTo": "JFK",
          "airline": "DY",
          "fare_classes": "",
          "cityFrom": "Oslo",
          "last_seen": "2019-02-15T12:58:48.000Z",
          "guarantee": false,
          "fare_basis": "",
          "local_arrival": "2019-03-01T20:00:00.000Z",
          "utc_arrival": "2019-03-02T01:00:00.000Z",
          "local_departure": "2019-03-01T17:45:00.000Z",
          "utc_departure": "2019-03-01T16:45:00.000Z"
        },
        {
          "refresh_timestamp": "2019-02-15T14:15:41.000Z",
          "bags_recheck_required": true,
          "return": 0,
          "flight_no": 2244,
          "operating_carrier": "DL",
          "cityTo": "Orlando",
          "vehicle_type": "aircraft",
          "flyFrom": "JFK",
          "id": "16669696135344215_0",
          "equipment": "321",
          "combination_id": "16669696135344215",
          "fare_family": "",
          "flyTo": "MCO",
          "airline": "DL",
          "fare_classes": "E",
          "cityFrom": "New York",
          "last_seen": "2019-02-15T14:15:41.000Z",
          "guarantee": true,
          "fare_basis": "VAUSA0BP",
          "local_arrival": "2019-03-02T11:01:00.000Z",
          "utc_arrival": "2019-03-02T16:01:00.000Z",
          "local_departure": "2019-03-02T08:05:00.000Z",
          "utc_departure": "2019-03-02T13:05:00.000Z"
        },
        {
          "refresh_timestamp": "1970-01-01T00:00:00.000Z",
          "bags_recheck_required": true,
          "return": 0,
          "flight_no": 1921,
          "operating_carrier": null,
          "cityTo": "Bogotá",
          "vehicle_type": "aircraft",
          "flyFrom": "MCO",
          "id": "10000058272100155_0",
          "equipment": null,
          "combination_id": "10000058272100155",
          "fare_family": "",
          "flyTo": "BOG",
          "airline": "NK",
          "fare_classes": "U",
          "cityFrom": "Orlando",
          "last_seen": "2019-02-15T08:09:42.000Z",
          "guarantee": true,
          "fare_basis": "UA14NR",
          "local_arrival": "2019-03-03T00:15:00.000Z",
          "utc_arrival": "2019-03-03T05:15:00.000Z",
          "local_departure": "2019-03-02T19:55:00.000Z",
          "utc_departure": "2019-03-03T00:55:00.000Z"
        },
        {
          "refresh_timestamp": "1970-01-01T00:00:00.000Z",
          "bags_recheck_required": true,
          "return": 0,
          "flight_no": 328,
          "operating_carrier": null,
          "cityTo": "Lima",
          "vehicle_type": "aircraft",
          "flyFrom": "BOG",
          "id": "19795607481891218_0",
          "equipment": null,
          "combination_id": "19795607481891218",
          "fare_family": "",
          "flyTo": "LIM",
          "airline": "VH",
          "fare_classes": "",
          "cityFrom": "Bogotá",
          "last_seen": "2019-02-15T12:54:00.000Z",
          "guarantee": true,
          "fare_basis": "",
          "local_arrival": "2019-03-03T12:11:00.000Z",
          "utc_arrival": "2019-03-03T17:11:00.000Z",
          "local_departure": "2019-03-03T09:00:00.000Z",
          "utc_departure": "2019-03-03T14:00:00.000Z"
        }
      ],
      "baggage": {
        "hold": [
          {
            "weight": 18,
            "price": 165.65,
            "dimensions_sum": 156,
            "height": 52,
            "width": 26,
            "length": 78,
            "tier": "1"
          },
          {
            "weight": 36,
            "price": 398.56,
            "dimensions_sum": 156,
            "height": 52,
            "width": 26,
            "length": 78,
            "tier": "3"
          },
          {
            "weight": 20,
            "price": 226.85,
            "dimensions_sum": 156,
            "height": 52,
            "width": 26,
            "length": 78,
            "tier": "2"
          }
        ],
        "personal_item": {
          "width": 10,
          "price": 0,
          "length": 33,
          "weight": 3,
          "height": 25
        }
      },
      "facilitated_booking_available": false,
      "countryFrom": {
        "code": "NO",
        "name": "Norway"
      },
      "local_arrival": "2019-03-03T12:11:00.000Z",
      "utc_arrival": "2019-03-03T17:11:00.000Z",
      "local_departure": "2019-03-01T17:45:00.000Z",
      "utc_departure": "2019-03-01T16:45:00.000Z"
    }
  ],
  "search_id": "365557ea-6145-41da-8789-7f0f78a02453"
}
