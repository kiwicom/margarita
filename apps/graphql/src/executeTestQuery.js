// @flow

import { graphql as originalGraphQL } from 'graphql';

import schema from './Schema';
import createContext from './services/GraphQLContext';

/**
 * Executes GraphQL query using our schema. This function should
 * be used only during testing because it uses faked context.
 */
export default function executeTestQuery(
  query: string,
  variables: ?Object,
  acceptLanguage: ?string,
): Promise<Object> {
  return originalGraphQL(schema, query, null, createContext(), variables);
}
