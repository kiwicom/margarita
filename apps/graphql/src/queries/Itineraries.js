// @flow

import { GraphQLNonNull } from 'graphql';
import { type ConnectionArguments } from 'graphql-relay';
import {
  connectionFromArray,
  connectionArgs,
  connectionDefinitions,
} from '@kiwicom/graphql-utils';

import ItinerariesSearchInput from '../types/input/ItinerariesSearchInput';
import GraphQLItinerary from '../types/output/Itinerary';
import type { GraphqlContextType } from '../services/GraphQLContext';
import type {
  ItinerariesType,
  ItinerariesSearchParametersType,
} from '../dataloaders/ItinerariesloaderTypes';

const { connectionType: ItinerariesConnection } = connectionDefinitions({
  nodeType: GraphQLItinerary,
});

type Args = {|
  +input: ItinerariesSearchParametersType,
  ...$Exact<ConnectionArguments>,
|};

export default {
  name: 'SearchItineraries',
  description: 'Query to fetch all itineraries(flights, busses, trains)',
  args: {
    input: {
      type: GraphQLNonNull(ItinerariesSearchInput),
    },
    ...connectionArgs,
  },
  type: ItinerariesConnection,
  resolve: async (_: mixed, args: Args, { dataLoader }: GraphqlContextType) => {
    const itineraries = await dataLoader.itineraries.load(args.input);

    return connectionFromArray<ItinerariesType>(itineraries, args);
  },
};
