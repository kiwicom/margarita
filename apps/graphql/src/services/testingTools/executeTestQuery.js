// @flow

import { graphql as originalGraphQL } from 'graphql';

import schema from '../../Schema';
import createContext from '../graphqlContext/GraphQLContext';

/**
 * Executes GraphQL query using our schema. This function should
 * be used only during testing because it uses faked context.
 */
export default function executeTestQuery(
  query: string,
  variables: ?Object,
): Promise<Object> {
  return originalGraphQL(schema, query, null, createContext(), variables);
}
