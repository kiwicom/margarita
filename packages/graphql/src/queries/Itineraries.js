// @flow

import { GraphQLNonNull } from 'graphql';
import { type ConnectionArguments } from 'graphql-relay';
import {
  connectionFromArray,
  connectionArgs,
  connectionDefinitions,
} from '@kiwicom/graphql-utils';

import ItinerariesSearchInput from '../types/input/ItinerariesSearchInput';
import GraphQLItinerary from '../types/output/Itinerary';
import type { GraphqlContextType } from '../services/GraphQLContext';
import type {
  Itineraries,
  ItinerariesSearchParameters,
} from '../dataloaders/Itinerariesloader';

const { connectionType: ItinerariesConnection } = connectionDefinitions({
  nodeType: GraphQLItinerary,
});

type Args = {|
  +input: ItinerariesSearchParameters,
  ...$Exact<ConnectionArguments>,
|};

export default {
  name: 'SearchItineraries',
  description: 'Query to fetch all itineraries(flights, busses, trains)',
  args: {
    input: {
      type: GraphQLNonNull(ItinerariesSearchInput),
    },
    ...connectionArgs,
  },
  type: ItinerariesConnection,
  resolve: async (_: mixed, args: Args, { dataLoader }: GraphqlContextType) => {
    const itineraries = await dataLoader.itineraries.load(args.input);

    return connectionFromArray<Itineraries>(itineraries, args);
  },
};
