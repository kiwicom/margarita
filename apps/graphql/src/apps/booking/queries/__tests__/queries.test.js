// @flow

import { generateTestsFromFixtures } from '@kiwicom/test-utils';

import executeTestQuery from '../../../../services/testingTools/executeTestQuery';
import Itineraries from '../../../itinerary/__datasets__/Itineraries.json';

fetch.mockResponses([JSON.stringify(Itineraries)]);

generateTestsFromFixtures(`${__dirname}/__fixtures__`, input =>
  executeTestQuery(input),
);
