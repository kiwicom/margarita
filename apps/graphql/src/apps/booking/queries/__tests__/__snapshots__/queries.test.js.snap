// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches expected output: BookingDetail.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment RouteStopFragment on RouteStop {
  time {
    utc
    local
  }
}

{
  bookingDetail(id:"Q3VzdG9tZXJCb29raW5nOjE2NDYzNDQ3") {
    id
    status
    arrival {
      ...RouteStopFragment
    }
    departure {
      ...RouteStopFragment
    }
    destinationImageUrl
    passengerCount
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "data": {
    "bookingDetail": {
      "id": "Qm9va2luZ09uZVdheToxNjQ2MzQ0Nw==",
      "status": "closed",
      "arrival": {
        "time": {
          "utc": "2018-11-28T16:20:00.000Z",
          "local": "2018-11-28T16:20:00.000Z"
        }
      },
      "departure": {
        "time": {
          "utc": "2018-11-28T11:20:00.000Z",
          "local": "2018-11-28T12:20:00.000Z"
        }
      },
      "destinationImageUrl": "https://images.kiwi.com/photos/600x600/las-palmas_es.jpg",
      "passengerCount": 1
    }
  }
}
`;

exports[`matches expected output: CustomerBookings.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment RouteStopFragment on RouteStop {
  time {
    utc
    local
  }
}

fragment SectorFragment on Sector {
  duration
}

{
  customerBookings {
    edges {
      node {
        id
        status
        arrival {
          ...RouteStopFragment
        }
        departure {
          ...RouteStopFragment
        }
        destinationImageUrl
        passengerCount

        ... on BookingOneWay {
          sector {
            ...SectorFragment
          }
        }
        ... on BookingReturn {
          inbound {
            ...SectorFragment
          }
          outbound {
            ...SectorFragment
          }
        }
        ... on BookingMulticity {
          sectors {
            ...SectorFragment
          }
        }
      }
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "data": {
    "customerBookings": {
      "edges": [
        {
          "node": {
            "id": "Qm9va2luZ09uZVdheToxNjQ2MzQ0Nw==",
            "status": "closed",
            "arrival": {
              "time": {
                "utc": "2018-11-28T16:20:00.000Z",
                "local": "2018-11-28T16:20:00.000Z"
              }
            },
            "departure": {
              "time": {
                "utc": "2018-11-28T11:20:00.000Z",
                "local": "2018-11-28T12:20:00.000Z"
              }
            },
            "destinationImageUrl": "https://images.kiwi.com/photos/600x600/las-palmas_es.jpg",
            "passengerCount": 1,
            "sector": {
              "duration": 300
            }
          }
        },
        {
          "node": {
            "id": "Qm9va2luZ1JldHVybjo4MTQyODI4",
            "status": "confirmed",
            "arrival": {
              "time": {
                "utc": "2018-09-21T12:31:00.000Z",
                "local": "2018-09-21T10:01:00.000Z"
              }
            },
            "departure": {
              "time": {
                "utc": "2018-09-20T19:30:00.000Z",
                "local": "2018-09-20T21:30:00.000Z"
              }
            },
            "destinationImageUrl": "https://images.kiwi.com/photos/600x600/st-johns_nl_ca.jpg",
            "passengerCount": 2,
            "inbound": {
              "duration": 915
            },
            "outbound": {
              "duration": 1021
            }
          }
        },
        {
          "node": {
            "id": "Qm9va2luZ011bHRpY2l0eTo2Njc2NTI0",
            "status": "confirmed",
            "arrival": {
              "time": {
                "utc": "2018-08-25T11:55:00.000Z",
                "local": "2018-08-25T13:55:00.000Z"
              }
            },
            "departure": {
              "time": {
                "utc": "2018-07-16T09:50:00.000Z",
                "local": "2018-07-16T11:50:00.000Z"
              }
            },
            "destinationImageUrl": "https://images.kiwi.com/photos/600x600/bratislava_sk.jpg",
            "passengerCount": 2,
            "sectors": [
              {
                "duration": 230
              },
              {
                "duration": 240
              },
              {
                "duration": 590
              }
            ]
          }
        }
      ]
    }
  }
}
`;
